src/usb.c

```diff
@@ -0,0 +1,1128 @@
+  g_object_set_data_full (G_OBJECT (request), "parent-window", g_strdup (arg_parent_window), g_free);
+  g_object_set_data_full (G_OBJECT (request), "usb-description", g_steal_pointer (&usb_description), g_free);
+
+  impl_request = xdp_impl_request_proxy_new_sync (g_dbus_proxy_get_connection (G_DBUS_PROXY (impl)),
+                                                  G_DBUS_PROXY_FLAGS_NONE,
+                                                  g_dbus_proxy_get_name (G_DBUS_PROXY (impl)),
+                                                  request->id,
+                                                  NULL, &error);
+  if (!impl_request)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  request_set_impl_request (request, impl_request);
+  request_export (request, g_dbus_method_invocation_get_connection (invocation));
+
+  xdp_usb_complete_request_permission (object, invocation, request->id);
+
+  task = g_task_new (object, NULL, NULL, NULL);
+  g_task_set_task_data (task, g_object_ref (request), g_object_unref);
+  g_task_run_in_thread (task, handle_request_permission_in_thread);
+
+  return TRUE;
+}
+
+static gboolean
+handle_create_monitor (XdpUsb                *object,
+                       GDBusMethodInvocation *invocation,
+                       GVariant              *arg_options)
+{
+  g_autoptr(GError) error = NULL;
+  UsbSession *usb_session = NULL;
+  Session *session = NULL;
+  Usb *usb = (Usb *) object;
+
+  usb_session = usb_session_new (arg_options, invocation, usb->udev, usb->sessions, &error);
+  if (!usb_session)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  session = (Session *) usb_session;
+
+  if (!session_export (session, &error))
+    {
+      g_warning ("Failed to export session: %s", error->message);
+      session_close (session, FALSE);
+    }
+  else
+    {
+      g_debug ("CreateMonitor new session '%s'",  session->id);
+      session_register (session);
+    }
+
+  xdp_usb_complete_create_monitor (object, invocation, session->id);
+  return TRUE;
+}
+
+static gboolean
+handle_enumerate_devices (XdpUsb                *object,
+                          GDBusMethodInvocation *invocation,
+                          GVariant              *options)
+{
+  Usb *usb = (Usb *) object;
+  GVariantBuilder builder;
+  g_autoptr(XdpAppInfo) app_info = xdp_invocation_lookup_app_info_sync (invocation, NULL, NULL);
+  g_autoptr(AppUsbPermissions) permissions = NULL;
+  g_autoptr(GVariantIter) filter = NULL;
+  g_auto(GVariantDict) options_dict;
+
+  g_variant_dict_init (&options_dict, options);
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("a{sa{sv}}"));
+
+  filter = get_filter_from_options (&options_dict);
+
+  permissions = app_usb_permissions_for_app_info (app_info);
+  usb_enumerate_all_to_variant (usb, &builder, permissions, filter);
```

Comment:
Can you please clarify how USB devices are enumerated the first time an app is run? It looks to me that this code would run into a chicken-and-egg problem, where apps can't enumerate devices because they don't have permissions stored, and they don't have permissions because they can't enumerate devices.